component/discovery/discovery.go

```diff
@@ -2,20 +2,64 @@ package discovery
 
 import (
 	"context"
+	"fmt"
 	"sort"
 	"sync"
 	"time"
 
 	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/pkg/cluster"
 	"github.com/prometheus/prometheus/discovery"
 	"github.com/prometheus/prometheus/discovery/targetgroup"
 	"github.com/prometheus/prometheus/model/labels"
+	"github.com/rfratto/ckit/shard"
 )
 
+// Clustering determines whether to use clustering.
+type Clustering struct {
+	NodeUpdates bool `river:"node_updates,attr"`
+}
+
 // Target refers to a singular discovered endpoint found by a discovery
 // component.
 type Target map[string]string
 
+// DistributedTargets allows for sharing discovery targets when a Flow
+// component is run in a clustered deployment.
+type DistributedTargets struct {
+	node    cluster.Node
+	targets []Target
+}
+
+// NewDistributedTargets creates the abstraction that allows components to
+// dynamically shard targets between components.
+func NewDistributedTargets(s cluster.Node, t []Target) DistributedTargets {
+	return DistributedTargets{s, t}
+}
+
+// Get implements ClusteredTargets. ???
+func (t *DistributedTargets) Get() []Target {
+	if t.node == nil {
+		return t.targets
+	}
+
+	res := make([]Target, 0, len(t.targets)/len(t.node.Peers()))
+	for _, tgt := range t.targets {
+		peers, err := t.node.Lookup(shard.StringKey(tgt.Labels().String()), 1, shard.OpReadWrite)
+		if err != nil {
+			continue
```

Comment:
If the cluster is a failing state, then no targets will get picked up by any agent. 
Maybe the default behavior should be to assign to self if an error happens here.

Reply:
This will error out only if we request more peers than they're available, which will never happen. I've gone ahead and done that as it's sensible in the general case though.